### ISSUE TITLE
Number of threads comprehension question

### ISSUE THREAD

**Issue by daniel-stockhausen** [benoitc/gunicorn#1656]
Hey,

if I start gunicorn with a config-file including

```
workers = 1
threads = 1
```

and no/default setting for the worker type, I get the following output from htop:

<img width="1520" alt="htop" src="https://user-images.githubusercontent.com/1905359/33549741-d9a7ed9c-d8eb-11e7-9734-8549e9e49429.png">

So if I understand that image correctly, there is one sub-process with 3 threads. If I increase the workers-parameter I get multiple of those sub-processes with 3 threads each. I would expect there to be the multiple sub-processes, but why are there 3 threads each? Could every one of those threads accept an incoming connection? 

Thanks in advance,
Daniel

**Comment by tilgovi** [comment]
I would expect two threads with the threaded worker:

The main thread
  - heartbeat to master process to keep the worker alive
  - accept connections

The worker thread(s)
  - thread pool of size `threads`

However, setting `threads = 1` actually causes Gunicorn to use the `sync` worker unless you also specify `worker_class = 'gthread'`, so I would expect to see only one thread.

Could your application be spawning threads? I do not reproduce what you see with a simple example on MacOS.

**Comment by NicolasLM** [comment]
There is a small issue in the docs then, as it gives the impression that setting `threads=42` will result in `worker_class='gthread'`:
https://github.com/benoitc/gunicorn/blob/master/docs/source/design.rst#how-many-threads

**Comment by daniel-stockhausen** [comment]
Thanks for your answers :)

I'm using gunicorn in combination with Flask - I don't know if Flask spawns any threads, couldn't find something useful about that.

@tilgovi what is the simple example you used in trying to reproduce it? I would like to try that myself and see if it behaves differently

**Comment by tilgovi** [comment]
@NicolasLM that's correct. It's just that when you set `threads=1` it won't :-D.

**Comment by tilgovi** [comment]
@daniel-stockhausen in the examples folder of Gunicorn source is a simple echo app.

```shell
> gunicorn -k gthread --threads 1 echo:app
```

**Comment by daniel-stockhausen** [comment]
For testing purposes I used a Docker container containing Ubuntu 16.04. Now my config only includes a "workers = 2". When I changed the number of cores for the container, I found something interesting:

1 Core
<img width="1503" alt="1-core" src="https://user-images.githubusercontent.com/1905359/33664933-8d572c90-da95-11e7-8ed0-42956439dad0.png">

2 Cores
<img width="1517" alt="2-cores" src="https://user-images.githubusercontent.com/1905359/33664945-939ebadc-da95-11e7-8aca-6ea49924ada1.png">

3 Cores
<img width="1509" alt="3-cores" src="https://user-images.githubusercontent.com/1905359/33664951-99261cb6-da95-11e7-8c56-3f283a73981c.png">

4 Cores
<img width="1514" alt="4-cores" src="https://user-images.githubusercontent.com/1905359/33664964-9f056808-da95-11e7-8578-1ef364090a16.png">

So in the first case (1 core) it seems to be as it's supposed to: 1 parent process (arbiter?) with 2 sub-processes (the 2 workers).
But with every additional core an additional thread for every worker appears.

**Then I tried the echo example:** It's always the same as in the previous 1-Core-Example: no threads, just the two workers and their parent - so it's working as it's supposed to. This means gunicorn is not the problem, right?

**Comment by tilgovi** [comment]
Is it otherwise the same configuration with both applications?

**Comment by daniel-stockhausen** [comment]
Yes, both the 4 cases with my own application as the results I had with the echo example had the same configuration with "workers = 2", no settings for threads or worker-type

**Comment by NicolasLM** [comment]
@daniel-stockhausen I bet that your code uses a library that spawns threads in the background, defaulting to 1 per core. Without seeing your code or your dependencies it's hard to tell more.

**Comment by daniel-stockhausen** [comment]
Yes, it's probably something in scikit-learn which I'm mainly using in my application. So I have to do some digging to find the source of the thread-spawning. Thank you all for your help, now I know it's not gunicorn and also understand better how gunicorn works :)

**Comment by benoitc** [comment]
threads = 1 means 2 os processes, + 1 thread in the worker, so I guess it's normal. (1 os process is just some kind of thread after all. Threads are created / workers. when threads = T and workers=K then it means hat you will have K * T threads + K processes + 1 (the master).

**Comment by daniel-stockhausen** [comment]
I've found the problem. For anyone else who might run into a similiar issue sometime:

The threads were created by numpy. Setting `OMP_NUM_THREADS=1` solved my problem. This can be done on the command line before executing your python application:
`export OMP_NUM_THREADS=1`

or in your application, but it's important to run the command before numpy is imported:
`import os`
`os.environ['OMP_NUM_THREADS'] = '1'`

Other suggestions I found to be used in the same way were:
`MKL_NUM_THREADS=1` or
`NUMEXPR_NUM_THREADS=1`
but at least in my case they both did not seem to make a difference.
